File: optional.ts
class Optional<A> {
  private _isValid: boolean;
  private _value: A | undefined;

  constructor(value?: A) {
    this._isValid = typeof value !== 'undefined';
    this._value = value;
  }
  
  isValid(): boolean {
    return this._isValid;
  }
  
  value(): A {
    if (this._isValid) {
      return this._value as A;
    }
    throw new Error(`the Optional is none`);
  }
}

function composeOptional<A, B, C>(m1: (a: A) => Optional<B>, m2: (b: B) => Optional<C>): (c: A) => Optional<C> {
  return function (x: A) {
    const y = m1(x);
    if (!y.isValid()) {
      return new Optional<C>();
    }
    return m2(y.value());
  }
}

function safe_root(x: number): Optional<number> {
  if (x > 0) {
    return new Optional(Math.sqrt(x));
  } else {
    return new Optional();
  }
}

function safe_reciprocal(x: number): Optional<number> {
  if (x != 0.0) {
    return new Optional(1/x);
  } else {
    return new Optional();
  }
}

function safe_root_reciprocal(x: number): Optional<number> {
  return composeOptional(safe_reciprocal, safe_root)(x);
}

console.log(safe_root_reciprocal(2).value());

File: optional.cpp
#include <cmath>

using namespace std;

template<class A> class optional {
  bool _isValid;
  A _value;
  public:
    optional() : _isValid(false) {}
    optional(A v) : _isValid(true), _value(v) {}
    bool isValid() const { return _isValid; }
    A value() const { return _value; }
};

auto const compose = [](auto m1, auto m2) {
  return [m1, m2](auto x) {
    auto p1 = m1(x);
    if (p1.isValid()) {
      return m2(p1.value());
    } else {
      return p1;  
    }
  };
};

template<class A> optional<A> identity(A x) {
  return optional<A>{x};
}

optional<double> safe_root(double x) {
  if (x > 0) {
    return optional<double>{sqrt(x)};
  } else {
    return optional<double>{};
  }
}

optional<double> safe_reciprocal(double x) {
  if (x != 0.0) {
    return optional<double>{1/x};
  } else {
    return optional<double>{};
  }
}

optional<double> safe_root_reciprocal(double x) {
  return compose(safe_reciprocal, safe_root)(x);
}
